src/server/journal/serializer.cc

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/journal/serializer.h"
+
+#include "base/io_buf.h"
+#include "base/logging.h"
+#include "io/io.h"
+#include "server/common.h"
+#include "server/error.h"
+#include "server/journal/types.h"
+#include "server/main_service.h"
+#include "server/transaction.h"
+
+using namespace std;
+
+using nonstd::make_unexpected;
+
+// TODO: Stolen from rdb, unite in common utils
+
+#define SET_OR_RETURN(expr, dest)              \
+  do {                                         \
+    auto exp_val = (expr);                     \
+    if (!exp_val) {                            \
+      VLOG(1) << "Error while calling " #expr; \
+      return exp_val.error();                  \
+    }                                          \
+    dest = exp_val.value();                    \
+  } while (0)
+
+#define SET_OR_UNEXPECT(expr, dest)            \
+  {                                            \
+    auto exp_res = (expr);                     \
+    if (!exp_res)                              \
+      return make_unexpected(exp_res.error()); \
+    dest = std::move(exp_res.value());         \
+  }
+
+namespace dfly {
+
+JournalWriter::JournalWriter(io::Sink* sink, std::optional<DbIndex> dbid)
+    : sink_{sink}, cur_dbid_{dbid} {
+}
+
+error_code JournalWriter::WriteU8(uint8_t v) {
+  return sink_->Write(io::Bytes{&v, 1});
+}
+
+error_code JournalWriter::WriteU16(uint16_t v) {
+  uint8_t buf[2];
+  absl::big_endian::Store16(&buf, v);
+  return sink_->Write(io::Bytes{buf, sizeof(buf)});
+}
+
+error_code JournalWriter::WriteU64(uint64_t v) {
```

Comment:
how about writing something similar to SerializeLen in rdb save?
Than you dont need to assume anything about the values you write . This can prevent bugs in calling the incorrect function and also may reduce the data size

Reply:
Only for reducing data size.

I want to use SerializeLen from rdb, so I'll create a commons file for them

But in the future we should move all serialization logic into one folder